### ISSUE TITLE
Waker pipe sets CLOEXEC with F_SETFL when pipe2 is unavailable.

### ISSUE THREAD

**Issue by reiver-dev** [tokio-rs/mio#1092]
Version: 0.6.19

sys/unix/mod.rs
```rust
    // Use pipe2 for atomically setting O_CLOEXEC if we can, but otherwise
    // just fall back to using `pipe`.
    dlsym!(fn pipe2(*mut c_int, c_int) -> c_int);
```

However then mio sets flags using F_SETFL, which result in CLOEXEC missed.

```
epoll_create1(0)                        = 3
fcntl(3, F_GETFD)                       = 0
fcntl(3, F_SETFD, FD_CLOEXEC)           = 0
pipe([4, 5])                            = 0
fcntl(4, F_SETFL, O_RDONLY|O_NONBLOCK|O_CLOEXEC) = 0
fcntl(5, F_SETFL, O_RDONLY|O_NONBLOCK|O_CLOEXEC) = 0
...
fcntl(3, F_GETFD)                       = 0x1 (flags FD_CLOEXEC)
fcntl(4, F_GETFD)                       = 0
fcntl(5, F_GETFD)                       = 0
```
epoll fd has correct flag, pipe fds will be inherited after exec

(seems to be not an issue in 0.7 due to use of eventfd)

**Comment by Thomasdezeeuw** [comment]
Can you provide some code that show how you got these results? Because I can't find the code that resets the flags in Mio.

**Comment by reiver-dev** [comment]
branch: v0.6.x
path: sys/unix/mod.rs
line: 47

```rust
pub fn pipe() -> ::io::Result<(Io, Io)> {
    // Use pipe2 for atomically setting O_CLOEXEC if we can, but otherwise
    // just fall back to using `pipe`.
    dlsym!(fn pipe2(*mut c_int, c_int) -> c_int);

    let mut pipes = [0; 2];
    let flags = libc::O_NONBLOCK | libc::O_CLOEXEC;
    unsafe {
        match pipe2.get() {
            Some(pipe2_fn) => {
                cvt(pipe2_fn(pipes.as_mut_ptr(), flags))?;
            }
            None => {
                cvt(libc::pipe(pipes.as_mut_ptr()))?;
                libc::fcntl(pipes[0], libc::F_SETFL, flags);
                libc::fcntl(pipes[1], libc::F_SETFL, flags);
            }
        }
    }

    unsafe {
        Ok((Io::from_raw_fd(pipes[0]), Io::from_raw_fd(pipes[1])))
    }
}
```

**Comment by Thomasdezeeuw** [comment]
I got that, but it sets both the `O_CLOEXEC` flag in both cases, so I don't see the problem.

**Comment by reiver-dev** [comment]
F_SETFL does not accept O_CLOEXEC on linux
https://github.com/tokio-rs/mio/pull/596

**Comment by Thomasdezeeuw** [comment]
Ah now I understand. Bit silly that Linux doesn't complain about anything and mixes `F_` constants with `O_`, but fine. I'll send a pr.

**Comment by Thomasdezeeuw** [comment]
This should have been closed by #1095.
